content/blog/2020-12-19-asmm/index.md

```diff
@@ -0,0 +1,119 @@
+
+## Current Limitations on Programs
+
+Outside of the type system additions, there are a few restriction put onto Bril Programs for them to be valid for this extension:
+
+- Any variable that owns it's memory(returned from ```alloc```) can not be overwritten during the course of the program. This is to ensure that a program doesn't make multiple allocations using the same variable name but whereas the static analysis can only free the variables once.
+- Currently, programmers cannot declare smaller regions within the function scope. Currently, each function has a region and each pointer argument has it's own region. The next step would be to allow the programmer to create their own scoped regions. This was set asider during the project proposal process but would be interesting further work.
+- As a consequence of the first two, a programmer cannot currently allocate memory in a loop. The ideal way to do this would to have the loop body be one region so that at the end of each loop iteration any memory allocated during that iteration is freed. I attempted to infer small regions just for this use case and found it to be both difficult infer and confusing to use. It can be very unclear where the analysis has chosen to start and end a region which makes for a pour experience when the programmer can't tell when a memory location goes out of scope.
+- Every Owning pointer must dominate the exit block(s) of it's region. In this way, you cannot conditionally allocate or free memory in a Bril Program. A given allocation of memory must exist in the same variable no matter which block it exits the region from. A borrowed pointer can be freely used as long as it is invalidated once the region scope ends.
+- The programmer can specify when two arguments to a function should have the same region but they are unable to declare one region as nested within another. This would allow finer grain control over what input regions are acceptable for a function can instead the coarseness of, "Either they are the same region or they aren't".
+- A pointer returned from a ```Call``` value instruction is assumed to own the memory it is returning. This is to make a clear distinction of who's job it is to free memory so that different functions don't double free the same memory at the end of their regions. This is more of an implementation limitation to make inference easier than a technical limitation.
```

Comment:
I dunno, I think this is not just an implementation limitation… it is pretty fundamental to shift from an intraprocedural analysis to an interprocedural one.

Reply:
Well, an implementation limitation in the sense that I can just add another pass after the intraprocedural one that all it does is check for this case. I didn't consider this to be that difficult, though I decided to avoid this to have a clear story of who frees memory and to hopefully have a stronger case for correctness. Being able to return a pointer that "borrows" memory instead of owning it seems to introduce a lot of possible bugs I need to check for.